### PR TITLE
introduce alternative WebSocket connectivity management

### LINKED ISSUE #147: [v1.0-beta] alternative WebSocket connectivity management
Managing websocket connectivity with ServerConfig  is not good.
It is really complicated to handle every usecase with mapping it to a exclusive state, and really buggy now... (#117, #137, #142, ...)

Let's introduce clean architecture partially just for managing the WebSocket connectivities.

---
slice

### PR DESCRIPTION
**Pull request by YusukeIwaki**
@RocketChat/android

Closes #147 

### ServerConfig is removed! 

* Now `ConnectivityManager` is responsible for the connectivity management.
* `serverConfigId` is no longer used, `hostname` is used as a primary key for each model.

### DDPClient#isConnected is removed!

`DDPClient#isConnected` returns true often even after connection is disconnected. It was really confusing! ( #137 is caused by it)
Now the function for checking connectivity is implemented with DDP ping-pong. It costs a little time, however it brings much more precise connectivity check.

---

### about RxJava usage...

I couldn't find `flatMap` or its compatible operator in `rx.Completable` .
( `andThen` looks similar but it doesn't evaluate item with laziness... ) So I used `Single<Boolean>` instead.

### CODE REVIEW

File: app/src/main/java/chat/rocket/android/helper/RxHelper.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ */
+public class RxHelper {
+  public static Func1<Observable<? extends Throwable>, Observable<?>> exponentialBackoff(
+      int maxRetryCount, long base, TimeUnit unit) {
+    return errors -> errors
+        .zipWith(Observable.range(0, maxRetryCount), (error, retryCount) -> retryCount)
+        .flatMap(retryCount -> Observable.timer(base * (long) Math.pow(2, retryCount), unit));
+  }
+
+  public static <T> Single<T> lazy(Func0<Single<T>> func) {
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/helper/RxHelper.java]: Won't `rx.Defer` be enough?

File: app/src/main/java/chat/rocket/android/helper/RxHelper.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import rx.functions.Func1;
+
+/**
+ * Rx operator and so on.
+ */
+public class RxHelper {
+  public static Func1<Observable<? extends Throwable>, Observable<?>> exponentialBackoff(
+      int maxRetryCount, long base, TimeUnit unit) {
+    return errors -> errors
+        .zipWith(Observable.range(0, maxRetryCount), (error, retryCount) -> retryCount)
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/helper/RxHelper.java:16]: Why `zipWith` if you're ignoring the `error`?

File: app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java [hunk 3]
```diff
@@ -123,111 +137,149 @@ private void forceInvalidateTokens() {
...
+      super.quit();
+      return Single.just(true);
     }
   }
 
   /**
+   * THIS METHOD THROWS EXCEPTION!!
+   * Use terminate() instead!!
+   */
+  @Deprecated public final boolean quit() {
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java]: `@Override`?

File: app/src/main/java/chat/rocket/android/service/ServerInfoImpl.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import chat.rocket.android.realm_helper.RealmHelper;
+import chat.rocket.android.realm_helper.RealmStore;
+
+/**
+ * Backend implementation to store ServerInfo.
+ */
+public class ServerInfoImpl extends RealmObject {
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/service/ServerInfoImpl.java]: This is not an implementation of `ServerInfo`... but a `Realm` version of it. Maybe `RealmServerInfo`?

File: app/src/main/java/chat/rocket/android/service/ConnectivityManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+import chat.rocket.android.log.RCLog;
+import hugo.weaving.DebugLog;
+import rx.Observable;
+import rx.Single;
+import rx.subjects.PublishSubject;
+
+/**
+ * Connectivity management implementation.
+ */
+/*package*/ class ConnectivityManagerImpl implements ConnectivityManagerApi, ConnectivityManagerInternal {
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/service/ConnectivityManagerImpl.java]: This `ConnectivityManagerImpl` talks directly with `ServerInfoImpl` (which is really a `RealmServerInfo`).

So maybe `RealmConnectivityManager` would be a better name.

File: app/src/main/java/chat/rocket/android/LaunchUtil.java [hunk 6]
```diff
@@ -23,10 +33,10 @@ public static void showAddServerActivity(Context context) {
   /**
    * launch ServerConfigActivity with proper flags.
    */
-  public static void showLoginActivity(Context context, String serverCondigId) {
+  public static void showLoginActivity(Context context, String hostname) {
     Intent intent = new Intent(context, LoginActivity.class);
     intent.setFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT | Intent.FLAG_ACTIVITY_CLEAR_TOP);
-    intent.putExtra("serverConfigId", serverCondigId);
+    intent.putExtra("hostname", hostname);
```
Comment (laggedHero) [inline app/src/main/java/chat/rocket/android/LaunchUtil.java]: What about a constant on `LaunchUtil`?

**Review by laggedHero — COMMENTED** [review]
As a rule of thumb `Impl` should be avoided: it usually means we have some bad design at hand.

File: app/src/main/java/chat/rocket/android/helper/RxHelper.java (hunk 2 above)
Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/helper/RxHelper.java:16]: Actually I made this using [javadoc](https://github.com/ReactiveX/RxJava/blob/a8ba158839b67246a742b6f1531995ffd7545c08/src/main/java/io/reactivex/Observable.java#L9601) as reference, and didn't know the detail...
I will study it...

File: app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java (hunk 3 above)
Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java]: This method is really not intended to use.
So I want to add some kind of annotation for indicating it.

File: app/src/main/java/chat/rocket/android/service/ServerInfoImpl.java (hunk 4 above)
Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/service/ServerInfoImpl.java]: Yes, exactry!
That sounds nice. I will fix.
Thank you.

File: app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java (hunk 3 above)
Comment (laggedHero) [reply app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java]: I understood that. But without the `@Override` it looks like you create a method just to deprecated it.

File: app/src/main/java/chat/rocket/android/helper/RxHelper.java (hunk 1 above)
Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/helper/RxHelper.java]: Oh, I didn't know that operator. Thank you.
`Single.defer` works as I expected.

File: app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java (hunk 3 above)
Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java]: > But without the @Override it looks like you create a method just to deprecated it.

Hmm, right. I will add `@Override`.

File: app/src/main/java/chat/rocket/android/helper/RxHelper.java (hunk 2 above)
Comment (laggedHero) [reply app/src/main/java/chat/rocket/android/helper/RxHelper.java:16]: Can't you just `flatMap` this?

Comment (YusukeIwaki) [reply app/src/main/java/chat/rocket/android/helper/RxHelper.java:16]: I think an operator like `enumerate()` is needed here.

slice

---

https://github.com/ReactiveX/RxJava/issues/3602

It seems that an operator for getting the index of each item is implemeted in `rxjava-extras`. However it also uses `.zipWith(Observable.range(0, maxRetryCount), (error, retryCount) -> retryCount)`. https://github.com/davidmoten/rxjava-extras/blob/a70df4359a818b240b053cb8bd8d6ce2a56109e9/src/main/java/com/github/davidmoten/rx/util/MapWithIndex.java#L25

I think this is reasonable. However maybe the name of variable `errors` is not good. I will change the variable name.

Comment (laggedHero) [reply app/src/main/java/chat/rocket/android/helper/RxHelper.java:16]: Yep. Sounds good. 👍

**Comment by laggedHero** [comment]
@YusukeIwaki planning more changes? I'd like to give it a spin with our alpha testers. xD

**Comment by YusukeIwaki** [comment]
@laggedHero No, this is all I wanted to try first.
If some bugs remain, I want fix them with other PRs.
Thank you for your reviewing and good advices!

**Review by laggedHero — APPROVED** [review]

**Pull request merged**